src/t8.h

```diff
@@ -73,6 +73,25 @@ T8_EXTERN_C_BEGIN ();
 #define T8_ASSERT(c) SC_NOOP ()
 #endif
 
+/** Extended assertion with detailed error message (including the condition, file, function, 
+ * and line number) along with a custom user-specified message. Only active in debug-mode.
+ * \param cond     The condition to check.
+ * \param msg      A custom message to display if the assertion fails.
+ * \param ...      Optional additional arguments to format the message.
+  */
+#ifdef T8_ENABLE_DEBUG
+#define T8_ASSERTF(cond, msg, ...) \
+  do { \
+    if (!(cond)) { \
+      fprintf (stderr, "Assertion failed: (%s), function %s, file %s, line %d.\nMessage: " msg "\n", #cond, \
+               __FUNCTION__, __FILE__, __LINE__, ##__VA_ARGS__); \
+      abort (); \
+    } \
+  } while (0)
```

Comment:
I do not like the while(0) approach here. Why do we need it? Can we remove it?

Reply:
As I understand it, this ensures that the macro always behaves as intended, for example, when used with external `if` statements. I got the explanation from [here](https://hownot2code.wordpress.com/2016/12/05/do-while-0-in-macros/).

Reply:
What about a t8code wrapper around `SC_CHECK_ABORT` ? It will provide the desired feature and also take care that t8code stops on all processes, not only the one that just failed.

Reply:
You are correct. The `SC_CHECK_ABORT`, which is currently used in `T8_ASSERT`, already has the capability to display an error message. Therefore, I have now added the option to include the error message directly in `T8_ASSERT`. As a result, there won't be an additional `T8_ASSERTF`, instead, you can now add the message directly in `T8_ASSERT`.

Reply:
Never mind, I added `T8_ASSERTF`, which displays a custom error message. `T8_ASSERT` itself stays untouched.